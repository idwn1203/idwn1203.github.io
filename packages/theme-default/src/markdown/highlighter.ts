import type { BundledLanguage } from 'shiki/langs'
import type { BundledTheme, CodeToHastOptions } from 'shiki'
import { getHighlighter } from 'shiki'
import type { Highlighter } from '@sveltepress/vite'
import { LRUCache } from 'lru-cache'
import { createTransformerFactory } from '@shikijs/twoslash'
import { createTwoslasher, rendererFloatingSvelte } from '@sveltepress/twoslash'
import { themeOptionsRef } from '../index.js'
import { processCommands } from './commands.js'

const DEFAULT_SUPPORT_LANGUAGES: BundledLanguage[] = ['svelte', 'sh', 'js', 'html', 'ts', 'md', 'css', 'scss']

const cache = new LRUCache<string, any>({ max: 1024 })

async function createHighlighterWithThemeAndLangs() {
  const highlighterConfig = themeOptionsRef.value?.highlighter
  const langs = highlighterConfig?.languages || DEFAULT_SUPPORT_LANGUAGES
  const darkTheme = highlighterConfig?.themeDark ?? 'night-owl'
  const lightTheme = highlighterConfig?.themeLight ?? 'vitesse-light'
  const shikiHighlighter = await getHighlighter({
    themes: [darkTheme, lightTheme],
    langs,
  })

  const _highlighter: Highlighter = async (code, lang) => {
    const shikiOptions: CodeToHastOptions<BundledLanguage, BundledTheme> = {
      lang,
      themes: {
        dark: darkTheme,
        light: lightTheme,
      },
      transformers: [],
    }
    if (themeOptionsRef.value?.highlighter?.twoslash) {
      shikiOptions.transformers?.push(createTransformerFactory(await createTwoslasher())({
        langs: ['ts', 'tsx', 'svelte'],
        renderer: rendererFloatingSvelte({
          lang: 'ts',
        }),
      }))
    }
    return shikiHighlighter.codeToHtml(code, shikiOptions)
      .replace(/\{/g, '&#123;')
      .replace(/\}/g, '&#125;')
  }

  return _highlighter
}

const highlighterRef: {
  value?: Highlighter
} = {}

async function ensureHighlighter() {
  if (!highlighterRef.value)
    highlighterRef.value = await createHighlighterWithThemeAndLangs()
}

const highlighter: Highlighter = async (code, lang, meta) => {
  const cacheKey = JSON.stringify({ code, lang, meta })
  let cached = cache.get(cacheKey)
  if (cached)
    return cached
  await ensureHighlighter()
  const metaArray = (meta || '').split(' ')
  const containLineNumbers = metaArray.some(item => item.trim() === 'ln')
  const titleMeta = metaArray.find(item => item.startsWith('title='))
  const commandDoms: string[] = []
  const lines = code.split('\n')
  let noErrorsFirstLine: string | undefined
  if (lines[0] === '// @noErrors')
    noErrorsFirstLine = lines.shift()
  if (lang !== 'md') {
    code = lines.map((line, i) => {
      const [commandDomsInOneLine, newLine] = processCommands(line, i, lines.length)
      commandDoms.push(...commandDomsInOneLine)
      return newLine
    }).join('\n')
  }
  let title: string | undefined
  if (titleMeta)
    title = titleMeta.split('=')[1].replace(/(^")|("$)/g, '')
  if (noErrorsFirstLine)
    code = `${noErrorsFirstLine}\n${code}`

  cached = `
<div class="svp-code-block-wrapper">${title
? `<div class="svp-code-block--title">${title}</div>
`
: ''}
  <div class="svp-code-block${containLineNumbers ? ' svp-code-block--with-line-numbers' : ''}">
    ${commandDoms.join('\n')}
    <!-- svelte-ignore a11y-no-noninteractive-tabindex -->
    ${await highlighterRef.value?.(code, lang, meta)}
    <div class="svp-code-block--lang">
      ${lang}
    </div>
    <CopyCode />
    ${containLineNumbers ? `<div class="svp-code-block--line-numbers">${lines.map((_, i) => `<div class="svp-code-block--line-number-item">${i + 1}</div>`).join('\n')}</div>` : ''}
  </div>
</div>`

  cache.set(cacheKey, cached)
  return cached
}

export default highlighter
