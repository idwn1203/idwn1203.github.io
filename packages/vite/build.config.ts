import { defineBuildConfig } from 'unbuild'

export default defineBuildConfig({
  rollup: {
    resolve: {},
    inlineDependencies: true,
  },
  failOnWarn: false,
})
