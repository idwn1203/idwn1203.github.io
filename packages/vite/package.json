{
  "name": "@sveltepress/vite",
  "homepage": "https://sveltepress.site/",
  "version": "0.33.2",
  "description": "Sveltepress plugin for vite",
  "files": [
    "dist",
    "types.d.ts"
  ],
  "type": "module",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": "./dist/index.mjs",
    "./types": {
      "types": "./types.d.ts"
    }
  },
  "scripts": {
    "test": "vitest",
    "build": "unbuild",
    "prepare": "pnpm build"
  },
  "keywords": [
    "vite",
    "vite plugin",
    "sveltepress",
    "sveltepress vite",
    "sveltekit"
  ],
  "author": {
    "name": "Dongsheng Zhao",
    "email": "1197160272@qq.com"
  },
  "repository": {
    "type": "github",
    "url": "https://github.com/Blackman99/sveltepress",
    "directory": "packages/vite"
  },
  "license": "MIT",
  "devDependencies": {
    "@sveltepress/theme-default": "workspace:*",
    "@sveltejs/kit": "^2.5.1",
    "@types/cross-spawn": "^6.0.6",
    "@types/estree": "^1.0.5",
    "@types/fs-extra": "^11.0.4",
    "@types/mdast": "^4.0.3",
    "@types/node": "^20.11.20",
    "dayjs": "^1.11.10",
    "svelte": "^4.2.12",
    "unbuild": "^2.0.0",
    "vite": "^5.1.4",
    "vitest": "^1.3.1"
  },
  "dependencies": {
    "cross-spawn": "^7.0.3",
    "fs-extra": "^11.2.0",
    "lru-cache": "^10.2.0",
    "picocolors": "^1.0.0",
    "rehype-stringify": "^10.0.0",
    "remark-directive": "^3.0.0",
    "remark-emoji": "^4.0.1",
    "remark-extract-frontmatter": "^3.2.0",
    "remark-frontmatter": "^5.0.0",
    "remark-gfm": "^4.0.0",
    "remark-parse": "^11.0.0",
    "remark-rehype": "^11.1.0",
    "remark-stringify": "^11.0.0",
    "shiki": "^1.1.7",
    "svelte-preprocess": "^5.1.3",
    "unified": "^11.0.4",
    "unist-util-visit": "^5.0.0",
    "vfile": "^6.0.1",
    "vite-plugin-inspect": "^0.8.3",
    "yaml": "^2.4.0"
  },
  "peerDependencies": {
    "@sveltejs/kit": "^1.20.4 || ^2.0.0",
    "@sveltejs/vite-plugin-svelte": "^3.0.1",
    "svelte": "^3.55.1 || ^4.0.0",
    "vite": "^4.3.9 || ^5.0.0"
  }
}
