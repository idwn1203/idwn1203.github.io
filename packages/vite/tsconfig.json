{
  "compilerOptions": {
    "allowJs": true,
    "checkJs": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "skipLibCheck": true,
    "strict": true,
    "declaration": true,
    "outDir": "dist",
    "moduleResolution": "node",
    "removeComments": false,
    "module": "ESNext",
  },
  "include": ["src", "types.d.ts"]
}