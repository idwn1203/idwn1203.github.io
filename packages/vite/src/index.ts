import type { PluginOption } from 'vite'
import { sveltekit } from '@sveltejs/kit/vite'
import vitePluginInspect from 'vite-plugin-inspect'
import SveltepressVitePlugin from './plugin.js'
import type { Highlighter, LoadTheme, ResolvedTheme, SiteConfig, SveltepressVitePluginOptions } from './types'
import mdToSvelte from './markdown/md-to-svelte'

export * as log from './utils/log.js'

const sveltepress: (options: SveltepressVitePluginOptions) => PluginOption = ({
  theme,
  addInspect,
  siteConfig,
  remarkPlugins,
  rehypePlugins,
} = {
  addInspect: false,
}) => {
  const requiredSiteConfig: Required<SiteConfig> = {
    title: siteConfig?.title || 'Untitled site',
    description: siteConfig?.description || 'Build by Sveltepress',
  }
  const corePlugin = [
    SveltepressVitePlugin({
      theme,
      siteConfig: requiredSiteConfig,
      remarkPlugins,
      rehypePlugins,
    }),
    sveltekit(),
  ]

  const plugins = typeof theme?.vitePlugins === 'function'
    ? theme.vitePlugins(corePlugin)
    : [
        theme?.vitePlugins,
        ...corePlugin,
      ]
  if (addInspect)
    plugins.unshift(vitePluginInspect())

  return plugins
}

export { sveltepress, mdToSvelte }
export type { LoadTheme, Highlighter, SiteConfig, SveltepressVitePluginOptions, ResolvedTheme }
