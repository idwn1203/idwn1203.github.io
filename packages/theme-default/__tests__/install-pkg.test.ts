import { describe, expect, it } from 'vitest'
import { mdToSvelte } from '@sveltepress/vite'
import installPkg, { pkgRe } from '../src/markdown/install-pkg'
import highlighter from '../src/markdown/highlighter'

describe('install-pkg', () => {
  it('re', () => {
    expect(pkgRe.exec('@install-pkg(foo-bar)')).toMatchInlineSnapshot(`
      [
        "@install-pkg(foo-bar)",
        "foo-bar",
        undefined,
      ]
    `)
    expect(pkgRe.exec('@install-pkg(foo-bar,create)')).toMatchInlineSnapshot(`
      [
        "@install-pkg(foo-bar,create)",
        "foo-bar,create",
        undefined,
      ]
    `)
  })

  it('md', async () => {
    const source = '@install-pkg(@sveltepress/vite,create)'

    const { code } = await mdToSvelte({
      mdContent: source,
      filename: 'install-pkg.md',
      highlighter,
      remarkPlugins: [installPkg],
    })
    expect(code).toMatchSnapshot()
  })
})
