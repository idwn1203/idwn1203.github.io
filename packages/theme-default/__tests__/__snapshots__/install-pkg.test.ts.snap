// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`install-pkg > md 1`] = `
"<InstallPkg><div slot="npm">
<div class="svp-code-block-wrapper">
  <div class="svp-code-block">
    
    <!-- svelte-ignore a11y-no-noninteractive-tabindex -->
    <pre class="shiki shiki-themes vitesse-light night-owl" style="background-color:#ffffff;--shiki-dark-bg:#011627;color:#393a34;--shiki-dark:#d6deeb" tabindex="0"><code><span class="line"><span style="color:#59873A;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic">npm</span><span style="color:#B56959;--shiki-dark:#ECC48D"> create</span><span style="color:#B56959;--shiki-dark:#ECC48D"> @sveltepress/vite</span></span></code></pre>
    <div class="svp-code-block--lang">
      sh
    </div>
    <CopyCode />
    
  </div>
</div></div><div slot="yarn">
<div class="svp-code-block-wrapper">
  <div class="svp-code-block">
    
    <!-- svelte-ignore a11y-no-noninteractive-tabindex -->
    <pre class="shiki shiki-themes vitesse-light night-owl" style="background-color:#ffffff;--shiki-dark-bg:#011627;color:#393a34;--shiki-dark:#d6deeb" tabindex="0"><code><span class="line"><span style="color:#59873A;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic">yarn</span><span style="color:#B56959;--shiki-dark:#ECC48D"> create</span><span style="color:#B56959;--shiki-dark:#ECC48D"> @sveltepress/vite</span></span></code></pre>
    <div class="svp-code-block--lang">
      sh
    </div>
    <CopyCode />
    
  </div>
</div></div><div slot="pnpm">
<div class="svp-code-block-wrapper">
  <div class="svp-code-block">
    
    <!-- svelte-ignore a11y-no-noninteractive-tabindex -->
    <pre class="shiki shiki-themes vitesse-light night-owl" style="background-color:#ffffff;--shiki-dark-bg:#011627;color:#393a34;--shiki-dark:#d6deeb" tabindex="0"><code><span class="line"><span style="color:#59873A;--shiki-dark:#82AAFF;font-style:inherit;--shiki-dark-font-style:italic">pnpm</span><span style="color:#B56959;--shiki-dark:#ECC48D"> create</span><span style="color:#B56959;--shiki-dark:#ECC48D"> @sveltepress/vite</span></span></code></pre>
    <div class="svp-code-block--lang">
      sh
    </div>
    <CopyCode />
    
  </div>
</div></div></InstallPkg>"
`;
